Runner eventually runs out of file descriptors while trying to reconnect

```
Feb 18 19:18:42 i44pc31 java[9978]: 19:18:42.216 [Thread-0] INFO  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Connecting
Feb 18 19:18:48 i44pc31 java[9978]: 19:18:48.376 [Thread-0] WARN  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Failed to connect, retrying soon
Feb 18 19:18:58 i44pc31 java[9978]: 19:18:58.376 [Thread-0] INFO  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Connecting
Feb 18 19:19:04 i44pc31 java[9978]: 19:19:04.536 [Thread-0] WARN  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Failed to connect, retrying soon
Feb 18 19:19:14 i44pc31 java[9978]: 19:19:14.537 [Thread-0] INFO  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Connecting
Feb 18 19:19:20 i44pc31 java[9978]: 19:19:20.699 [Thread-0] WARN  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Failed to connect, retrying soon
Feb 18 19:19:30 i44pc31 java[9978]: 19:19:30.699 [Thread-0] INFO  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Connecting
Feb 18 19:19:36 i44pc31 java[9978]: 19:19:36.855 [Thread-0] WARN  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Failed to connect, retrying soon
Feb 18 19:19:46 i44pc31 java[9978]: 19:19:46.856 [Thread-0] INFO  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Connecting
Feb 18 19:19:53 i44pc31 java[9978]: 19:19:53.016 [Thread-0] WARN  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Failed to connect, retrying soon
Feb 18 19:20:03 i44pc31 java[9978]: 19:20:03.016 [Thread-0] INFO  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Connecting
Feb 18 19:20:09 i44pc31 java[9978]: 19:20:09.176 [Thread-0] WARN  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Failed to connect, retrying soon
Feb 18 19:20:19 i44pc31 java[9978]: 19:20:19.176 [Thread-0] INFO  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Connecting
Feb 18 19:20:25 i44pc31 java[9978]: 19:20:25.336 [Thread-0] WARN  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Failed to connect, retrying soon
Feb 18 19:20:35 i44pc31 java[9978]: 19:20:35.336 [Thread-0] INFO  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Connecting
Feb 18 19:20:41 i44pc31 java[9978]: 19:20:41.501 [Thread-0] WARN  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Failed to connect, retrying soon
Feb 18 19:20:51 i44pc31 java[9978]: 19:20:51.501 [Thread-0] INFO  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Connecting
Feb 18 19:20:57 i44pc31 java[9978]: 19:20:57.656 [Thread-0] WARN  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Failed to connect, retrying soon
Feb 18 19:21:07 i44pc31 java[9978]: 19:21:07.656 [Thread-0] INFO  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Connecting
Feb 18 19:21:13 i44pc31 java[9978]: 19:21:13.816 [Thread-0] WARN  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Failed to connect, retrying soon
Feb 18 19:21:23 i44pc31 java[9978]: 19:21:23.816 [Thread-0] INFO  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Connecting
Feb 18 19:21:29 i44pc31 java[9978]: 19:21:29.976 [Thread-0] WARN  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Failed to connect, retrying soon
Feb 18 19:21:39 i44pc31 java[9978]: 19:21:39.976 [Thread-0] INFO  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Connecting
Feb 18 19:21:46 i44pc31 java[9978]: 19:21:46.136 [Thread-0] WARN  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Failed to connect, retrying soon
Feb 18 19:21:56 i44pc31 java[9978]: 19:21:56.136 [Thread-0] INFO  d.aaaaaaah.velcom.runner.TeleBackend - ws://i44pc22:9666/runner-connector - Connecting
Feb 18 19:21:56 i44pc31 java[9978]: Exception in thread "Thread-0" java.lang.InternalError: java.io.IOException: Zu viele offene Dateien
Feb 18 19:21:56 i44pc31 java[9978]:         at java.net.http/jdk.internal.net.http.HttpClientImpl.<init>(HttpClientImpl.java:311)
Feb 18 19:21:56 i44pc31 java[9978]:         at java.net.http/jdk.internal.net.http.HttpClientImpl.create(HttpClientImpl.java:253)
Feb 18 19:21:56 i44pc31 java[9978]:         at java.net.http/jdk.internal.net.http.HttpClientBuilderImpl.build(HttpClientBuilderImpl.java:135)
Feb 18 19:21:56 i44pc31 java[9978]:         at java.net.http/java.net.http.HttpClient.newHttpClient(HttpClient.java:158)
Feb 18 19:21:56 i44pc31 java[9978]:         at de.aaaaaaah.velcom.runner.Connection.<init>(Connection.java:51)
Feb 18 19:21:56 i44pc31 java[9978]:         at de.aaaaaaah.velcom.runner.TeleBackend.run(TeleBackend.java:92)
Feb 18 19:21:56 i44pc31 java[9978]:         at de.aaaaaaah.velcom.runner.RunnerMain.lambda$realMain$1(RunnerMain.java:62)
Feb 18 19:21:56 i44pc31 java[9978]:         at java.base/java.lang.Thread.run(Thread.java:834)
Feb 18 19:21:56 i44pc31 java[9978]: Caused by: java.io.IOException: Zu viele offene Dateien
Feb 18 19:21:56 i44pc31 java[9978]:         at java.base/sun.nio.ch.IOUtil.makePipe(Native Method)
Feb 18 19:21:56 i44pc31 java[9978]:         at java.base/sun.nio.ch.EPollSelectorImpl.<init>(EPollSelectorImpl.java:83)
Feb 18 19:21:56 i44pc31 java[9978]:         at java.base/sun.nio.ch.EPollSelectorProvider.openSelector(EPollSelectorProvider.java:36)
Feb 18 19:21:56 i44pc31 java[9978]:         at java.base/java.nio.channels.Selector.open(Selector.java:295)
Feb 18 19:21:56 i44pc31 java[9978]:         at java.net.http/jdk.internal.net.http.HttpClientImpl$SelectorManager.<init>(HttpClientImpl.java:699)
Feb 18 19:21:56 i44pc31 java[9978]:         at java.net.http/jdk.internal.net.http.HttpClientImpl.<init>(HttpClientImpl.java:308)
Feb 18 19:21:56 i44pc31 java[9978]:         ... 7 more
```

--- a/backend/runner/src/main/java/de/aaaaaaah/velcom/runner/Connection.java
+++ b/backend/runner/src/main/java/de/aaaaaaah/velcom/runner/Connection.java
@@ -38,8 +38,8 @@ public class Connection implements WebSocket.Listener, HeartbeatWebsocket {
 	private WebSocket socket;
 	private HeartbeatHandler heartbeatHandler;
 
-	public Connection(TeleBackend teleBackend, URI address, String name, String token)
-		throws ExecutionException, InterruptedException {
+	public Connection(TeleBackend teleBackend, HttpClient httpClient, URI address, String name,
+		String token) throws ExecutionException, InterruptedException {
 
 		stateMachine = new StateMachine<>(new Idle(teleBackend, this));
 		serializer = new Serializer();
@@ -48,8 +48,7 @@ public class Connection implements WebSocket.Listener, HeartbeatWebsocket {
 		closed = false;
 
 		LOGGER.debug("Opening connection to {}", address);
-		HttpClient.newHttpClient()
-			.newWebSocketBuilder()
+		httpClient.newWebSocketBuilder()
 			.header(RunnerConnectionHeader.CONNECT_RUNNER_NAME.getName(), name)
 			.header(RunnerConnectionHeader.CONNECT_RUNNER_TOKEN.getName(), token)
 			.buildAsync(address, this)

--- a/backend/runner/src/main/java/de/aaaaaaah/velcom/runner/TeleBackend.java
+++ b/backend/runner/src/main/java/de/aaaaaaah/velcom/runner/TeleBackend.java
@@ -16,6 +16,7 @@ import de.aaaaaaah.velcom.shared.util.compression.TarHelper;
 import de.aaaaaaah.velcom.shared.util.systeminfo.LinuxSystemInfo;
 import java.io.IOException;
 import java.net.URI;
+import java.net.http.HttpClient;
 import java.nio.file.Path;
 import java.time.Instant;
 import java.util.Optional;
@@ -85,11 +86,20 @@ public class TeleBackend {
 	 * @throws InterruptedException if the thread is interrupted
 	 */
 	public void run() throws InterruptedException {
+		// Shared HttpClient to prevent FDs leaking based on GC
+		//
+		// A HttpClient initially creates/opens a few file descriptors. It only frees those file
+		// descriptors once no more references to the HttpClient exist. This usually requires garbage
+		// collection to take place. If a new HttpClient is created for each connection attempt but
+		// garbage collection only happens infrequently, file descriptors are "leaked". This can easily
+		// be prevented by using a single HttpClient instance for all connection attempts.
+		HttpClient httpClient = HttpClient.newHttpClient();
+
 		//noinspection InfiniteLoopStatement
 		while (true) {
 			try {
 				LOGGER.info("{} - Connecting", address);
-				Connection conn = new Connection(this, address, name, token);
+				Connection conn = new Connection(this, httpClient, address, name, token);
 				LOGGER.info("{} - Connected", address);
 				connection = conn;
 				conn.getClosedFuture().get();
@@ -97,7 +107,6 @@ public class TeleBackend {
 			} catch (ExecutionException | InterruptedException e) {
 				LOGGER.warn("{} - Failed to connect, retrying in {}", address,
 					Delays.RECONNECT_AFTER_FAILED_CONNECTION);
-				System.gc();
 				//noinspection BusyWait
 				Thread.sleep(Delays.RECONNECT_AFTER_FAILED_CONNECTION.toMillis());
 			}

--- a/backend/runner/src/main/java/de/aaaaaaah/velcom/runner/TeleBackend.java
+++ b/backend/runner/src/main/java/de/aaaaaaah/velcom/runner/TeleBackend.java
@@ -97,6 +97,7 @@ public class TeleBackend {
 			} catch (ExecutionException | InterruptedException e) {
 				LOGGER.warn("{} - Failed to connect, retrying in {}", address,
 					Delays.RECONNECT_AFTER_FAILED_CONNECTION);
+				System.gc();
 				//noinspection BusyWait
 				Thread.sleep(Delays.RECONNECT_AFTER_FAILED_CONNECTION.toMillis());
 			}

